ide/static/ide/js/ycm.js

```diff
@@ -1,4 +1,78 @@
 CloudPebble = CloudPebble || {};
+
+// This is a drop-in replacement for $.ajax
+// TODO: is this overkill? It seems to work, and means that very few changes to other code are required.
+// TODO: If we do keep this, (a) stick it in another file, (b) make sure it's robust
+var EventedWebSocket = function(host) {
+    var self = this;
+    var mSocket = null;
+    var ids = {};
+
+    _.extend(this, Backbone.Events);
+
+    this.getSocket = function() {
+        return mSocket;
+    };
+
+    this.connect = function() {
+        mSocket = new WebSocket(host);
+        mSocket.onerror = function(e) { self.trigger('error', e); };
+        mSocket.onclose = handle_socket_close;
+        mSocket.onmessage = handle_socket_message;
+        mSocket.onopen = function() { self.trigger('open'); };
+    };
+
+    this.close = function() {
+        mSocket.close();
+        mSocket = null;
+    };
+
+    this.send = function(data) {
+        var d = $.Deferred();
+        // Find the first free ID
+        var id;
+        for (id = 0; id in ids; id++);
+        // Register the promise
+        ids[id] = d;
+        data._ws_message_id = id;
+        // Send the data
+        var text = JSON.stringify(data);
+        mSocket.send(text);
+
+        return d.promise();
+    };
+
+    this.isOpen = function() {
+        return (mSocket.readyState == WebSocket.OPEN);
+    };
+
+
+    var handle_socket_close = function(e) {
+        self.trigger('close', e);
+        mSocket = null;
+        _.each(ids, function(promise, key) {
+            promise.reject({'error': 'closed'})
+        });
+        ids = {};
+    };
+
+    var handle_socket_message = function(e) {
+        var data = JSON.parse(e.data);
+        var id = data._ws_message_id;
+        if (id in ids) {
+            // Trigger the promise
+            var promise = ids[id];
+            delete ids[id];
+            if ('success' in data && data['success'] == false) {
```

Comment:
No good reason, I suppose! And yeah, I agree. But I think not having 'success' at all might be better, just to save on the extra bandwidth. If all errors come with error message, an 'error' key (or lack thereof) should be enough